src/main/include/log4cxx/log4cxx.h.in

```diff
@@ -76,8 +83,11 @@ typedef unsigned int log4cxx_uint32_t;
 //    definitions used when using DLL
 #define LOG4CXX_EXPORT __declspec(dllimport)
 #endif
+#define LOG4CXX_INSTANTIATE_EXPORTED_PTR(T) template class LOG4CXX_EXPORT std::shared_ptr<T>
 #else
 #define LOG4CXX_EXPORT
+#define LOG4CXX_DECLARE_PRIVATE_MEMBER_PTR(T, V) struct T; LOG4CXX_PRIVATE_PTR(T) V;
+#define LOG4CXX_INSTANTIATE_EXPORTED_PTR(T)
```

Comment:
Do we need to have two versions of this macro?  I don't know enough about how windows works, but would it be possible to just add `LOG4CXX_EXPORT` to the already-existing `LOG4CXX_PTR_DEF` macro?

Reply:
gcc says instantiated templates must be in their own namespace (not sure how to interpret that when the template and the template parameter are in different namespaces). Visual Studio does not issue that error, but not instantiating the template at all results in a  'needs to have a dll interface' warning.

My solution was to only instantiate when using Visual Studio.

See for example, in propertysetter.h